Ae question previews

andrewebdev opened the pull request:
## Description

Added a question preview component. The component provides a single preview
button and hidden content that form part of the preview.

When the preview button is clicked, the modal opens with the preview inside.

Fixes #224

## Type of change
Please delete options that are not relevant

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?
Tested manually by adding the component to the styleguide, with some complex
content for testing tab and focus behaviour.

Tested by opening and closing the preview with mouse and keyboard. I also tested
that the browser history back and forward buttons behave as expected on the
component.

Lastly I also tested that scrolling does the right thing; Ensure that the
content behind the modal does not scroll.

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I updated the CHANGELOG.md and added documentation if necessary
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have completed manual or automated accessibility testing for my changes
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

jupiter007 commented on app/[locale]/styleguide/page.tsx:
```diff
@@ -75,6 +75,8 @@ import {ModalOverlayComponent} from '@/components/ModalOverlayComponent';
 import ButtonWithImage from '@/components/ButtonWithImage';
 import {useToast} from '@/context/ToastContext';
 
+import QuestionPreview from '@/components/QuestionPreview';
```
L385 - Remember to add a link to the `Preview` demo in the Table of Contents

jupiter007 commented on components/QuestionPreview/index.tsx:
```diff
@@ -0,0 +1,125 @@
+const QuestionPreview: React.FC<QuestionPreviewProps> = ({
+  children,
+  id='',
+  className='',
```
L25: I don't see `className` being used anywhere in the component

jupiter007 commented on components/QuestionPreview/__tests__/index.spec.tsx:
```diff
@@ -0,0 +1,146 @@
+    });
+
+    window.history.pushState(null, "", window.location.pathname);
+    waitFor(() => {
```
Please add the axe accessibility unit test:
  ```
it('should pass axe accessibility test', async () => {

    const { container } = render(
      <QuestionPreview />
    );
    await act(async () => {
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });
```

andrewebdev replied:
This is actually something I wanted to chat about. I actually thought you added this as one of the tools similar to translations. Which is why I didn't add the test originally.

I added it now :+1:

jupiter007 commented on components/QuestionPreview/__tests__/index.spec.tsx:
```diff
@@ -0,0 +1,146 @@
+import {
+  render,
+  screen,
+  act,
```
L6: `act` is not used in the unit test
L12 & L13: Also, useParams, useRouter and useTranslations are not used in this test

`useTranslations` mock is configured in `jest.config.ts` so you don't have to include it in your test. Same goes for `useToast`. Also, I started a `__mocks__/common.ts` file to add some commonly used mocks for unit tests there as well.

jupiter007 reviewed (commented):
@andrewebdev the Preview component looks great. I stepped through it with `VoiceOver` and it looked good. Thanks for making the changes.

I only have a few comments.

andrewebdev commented:
@jupiter007 I addressed all the other issues. Regarding your last comment on adding the preview to template, we probably need more of a discussion. We changed how this component works from how it was originally designed (as a standalone page), so we need to revisit how we are going to do previews on Templates. It's probably worth a separate ticket.

We also need to update the question types ticket(s) (or create new tickets) to now add the preview for each question on the same page where the question is created.

jupiter007 commented:
> @jupiter007 I addressed all the other issues. Regarding your last comment on adding the preview to template, we probably need more of a discussion. We changed how this component works from how it was originally designed (as a standalone page), so we need to revisit how we are going to do previews on Templates. It's probably worth a separate ticket.
> 
> We also need to update the question types ticket(s) (or create new tickets) to now add the preview for each question on the same page where the question is created.

Thanks for making the updates @andrewebdev. I'll add an item on our agenda for next week about applying the preview to pages.
